v2/periphery/LlamaPayRouterOracle.sol

```diff
@@ -0,0 +1,807 @@
+
+/// @title LlamaPayRouterOracle.
+/// @notice An router/oracle that calculates the value of LlamaPay streams.
+///         The router/oracle can create, cancel, pause, modify, deposit and withdraw LlamaPay streams.
+/// @notice Note 1: The contract does not check for LlamaPay contract duplicates in constructor.
+/// @notice Note 2: All "amountPerSec" arguments should use 20 decimals.
+/// @notice Note 3: In the withdrawPayer() function, "amount" argument should use 20 decimals.
+/// @notice ### Audit Note: L-01 Vault Value Can Go Below Daily Lower Bound
+/// @notice Since a withdrawal action by the beneficiary of a LlamaPay instance does not affect the daily
+///         multiplier of a vault, the value locked within a vault over any 24-hour period can reach a lower
+///         value than would be expected, given the `minDailyValue`.
+///         Assume the following scenario:
+///         * The vault holds `X` amount of value and has a `minDailyValue` multiplier of `0.9`.
+///         * There is `Y = 0.03 * X` amount of value sitting in a LlamaPay contract managed by
+///         the `LlamaPayRouterOracle` contract.
+///         * The beneficiary of the LlamaPay stream withdraws `Y` value. Now there is `X - Y =
+///         0.97 * X` value in the vault, but the `cumulativeDailyMultiplier` stays at `1`.
+///         * The guardian submits a batch of actions which reduces the amount of value in the vault
+///         by `0.09 * X`. This would put the vault at `0.88 * X` value which is below the
+///         threshold of `0.9 * X`. Since the `cumulativeDailyMultiplier` did not account for
+///         the withdrawal of funds from the LlamaPay stream, the calculation of the
+///         `newMultiplier` in the `afterSubmit` hook results in a value greater than `0.9` and
+///         therefore allows it. The exact calculation of `newMultiplier`
+///         in this case is the following: `(1 * (0.88 * X)) / (0.97 * X)`,
+///         which results in `~0.907 > 0.9`. Therefore, the daily multiplier check passes.
+contract LlamaPayRouterOracle is
+    ILlamaPayRouterOracle,
+    AbstractAssetOracle,
+    Executor
+{
+    using SafeERC20 for IERC20;
+
+    /// CONSTANTS ///
+
+    /// @notice The maximum number of LlamaPay contracts.
+    uint256 private constant _MAX_LLAMAPAY_COUNT = 10;
+    /// @notice Downscale factor from LlamaPay balance to 18 decimals.
+    uint256 private constant _LLAMAPAY_DOWNSCALE_FACTOR = 1e2;
+    /// @notice The number of decimals of the LlamaPay payer balance.
+    uint256 private constant _LLAMAPAY_DECIMALS = 20;
+    /// @notice The name of the token.
+    string private constant _NAME = "LlamaPayRouterOracle";
+    /// @notice The symbol of the token.
+    string private constant _SYMBOL = "LPRO";
+
+    /// IMMUTABLES ///
+
+    /// @notice Number of LlamaPay contracts.
+    uint256 public immutable llamaPayCount;
+
+    /// @notice Decomposed list of LlamaPay contracts (one for each token).
```

Comment:
What is the design motivation behind unpacking `LlamaPayInfo[]`?  Also, If the uints were instead `uint96`, you could also pack storage slots to use 3 instead of 6 slots per `LlamaPayInfo`.

Reply:
Storing the instances as immutable variables reduces the number of `SLOADs` we need to do on oracle calls.